<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Joint Modeling (e.g., Multiomics) with Stray::Orthus • stray</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Joint Modeling (e.g., Multiomics) with Stray::Orthus">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stray</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.12</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://github.com/jsilve24/stray/wiki/Installation-Details">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction-to-stray.html">Intro to stray through stray::pibble</a>
    </li>
    <li>
      <a href="../articles/non-linear-models.html">Non-Linear Modeling with stray::basset</a>
    </li>
    <li>
      <a href="../articles/orthus.html">Joint Modeling  (e.g., Multiomics) with Stray::Orthus</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Other Packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://jsilve24.github.io/driver/">driver</a>
    </li>
    <li>
      <a href="https://bioconductor.org/packages/release/bioc/html/philr.html">philr</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/inschool4life">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/jsilve24/stray">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Joint Modeling (e.g., Multiomics) with Stray::Orthus</h1>
                        <h4 class="author">Justin Silverman</h4>
            
      
      
      <div class="hidden name"><code>orthus.Rmd</code></div>

    </div>

    
    
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(stray)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(phyloseq)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dplyr)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(driver)

<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Random">set.seed</a></span>(<span class="dv">48482</span>)</code></pre></div>
<p>If you have not already done so, I would read through the <em>pibble</em> vignette before this one.</p>
<div id="strayorthus-the-model" class="section level1">
<h1 class="hasAnchor">
<a href="#strayorthus-the-model" class="anchor"></a>stray::orthus, the model</h1>
<p>Stray can be used for jointly modeling multivariate count data and multivariate Gaussian data. For example, this would be a reasonable model to jointly model 16S microbiome data and metabolomics data jointly. Because of the “two-headed” nature of this model, e.g., two observed datasets, I named this model <em>orthus</em> (a two-headed dog and brother of Cerberus in Greek Mythology)[<a href="https://en.wikipedia.org/wiki/Orthrus" class="uri">https://en.wikipedia.org/wiki/Orthrus</a>]. The <em>orthus</em> model can be written as</p>
<p><span class="math display">\[
\begin{align}
Y_j &amp; \sim \text{Multinomial}\left(\pi_j \right)  \\
\pi_j &amp; = \phi^{-1}(\eta_j) \\
\begin{bmatrix}\eta_j \\ Z_j \end{bmatrix} &amp;\sim N(\Lambda X, \Sigma) \\
\Lambda &amp;\sim  N(\Theta, \Sigma, \Gamma) \\
\Sigma &amp;\sim W^{-1}(\Xi, \upsilon) 
\end{align}
\]</span></p>
<p>Note this looks nearly identical to the <em>pibble</em> model but we have appended the second (Gaussian) dataset (<span class="math inline">\(Z\)</span>) onto <span class="math inline">\(\eta\)</span>. In doing this, the definition of <span class="math inline">\(\Lambda\)</span> changes (it is now larger with the bottom rows dictating how the covariates <span class="math inline">\(X\)</span> influence the second dataset). Similarly, <span class="math inline">\(\Sigma\)</span> now is much larger and can be though of as <span class="math display">\[
\Sigma = \begin{bmatrix} \Sigma_{(\eta, \eta)} &amp; \Sigma_{(\eta, Z)} \\
                          \Sigma_{(Z, \eta)} &amp; \Sigma_{(Z, Z)}\end{bmatrix}
\]</span> where <span class="math inline">\(\Sigma_{(\eta, \eta)}\)</span> describes the covariance between log-ratios (e.g., the covariance among the multinomial categories in log-ratio space), <span class="math inline">\(\Sigma_{(Z, Z)}\)</span> describes the covariance between the dimensions of <span class="math inline">\(Z\)</span> (e.g., between metabolites if Z is metabolomics data), and <span class="math inline">\(\Sigma_{(\eta, Z)} = \Sigma_{(Z, \eta)}^T\)</span> represents the covariance between log-ratios and dimensions of <span class="math inline">\(Z\)</span> (e.g., between microbial taxa and metabolites). Similar to <span class="math inline">\(\Sigma\)</span> and <span class="math inline">\(\Lambda\)</span>, the parameters <span class="math inline">\(\Xi\)</span> and <span class="math inline">\(\Theta\)</span> undergo a similar expansion to accomidate the second dataset.</p>
</div>
<div id="joint-modeling-of-microbial-16s-data-and-metabolomics" class="section level1">
<h1 class="hasAnchor">
<a href="#joint-modeling-of-microbial-16s-data-and-metabolomics" class="anchor"></a>Joint modeling of Microbial 16S data and Metabolomics</h1>
<p>To demonstrate <em>orthus</em> I will perform a toy analysis on data from <span class="citation">Kashyap et al. (2013)</span> and made available by <span class="citation">Callahan et al. (2016)</span> as part of their recently published microbioem data anlaysis workflow <span class="citation">(Callahan et al. 2016)</span>. I follow the data preprocessing of <span class="citation">Callahan et al. (2016)</span> we just don’t drop taxa but instead amalgamate those that don’t pass filtering to a category called “other”. I do this to maintain the proper variance in the multinomial model.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">metab_path &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/Kashyap2013"</span>, <span class="st">"metabolites.csv"</span>, <span class="dt">package=</span><span class="st">"stray"</span>)
microbe_path &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/Kashyap2013"</span>, <span class="st">"microbe.rda"</span>, <span class="dt">package=</span><span class="st">"stray"</span>)
metab &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.csv</a></span>(metab_path, <span class="dt">row.names =</span> <span class="dv">1</span>)
metab &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>(metab)
microbe &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/get">get</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/load">load</a></span>(microbe_path))

## Preprocessing ##
<span class="co"># Metabolite Preprocessing</span>
keep_ix &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowSums</a></span>(metab <span class="op">==</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&lt;=</span><span class="st"> </span><span class="dv">3</span>
metab &lt;-<span class="st"> </span>metab[keep_ix, ]

<span class="co"># 16S Preprocesing - plus some weirdness to rename amalgamated category to "other"</span>
keep_ix &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/taxa_sums">taxa_sums</a></span>(microbe) <span class="op">&gt;</span><span class="st"> </span><span class="dv">4</span>
keep_ix &lt;-<span class="st"> </span>keep_ix <span class="op">&amp;</span><span class="st"> </span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowSums</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/otu_table-methods">otu_table</a></span>(microbe)<span class="op">&gt;</span><span class="dv">2</span>)<span class="op">&gt;</span><span class="dv">3</span>)
microbe &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/merge_taxa-methods">merge_taxa</a></span>(microbe, <span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/taxa_names-methods">taxa_names</a></span>(microbe)[<span class="op">!</span>keep_ix])
nms &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/taxa_names-methods">taxa_names</a></span>(microbe)
rnm &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/taxa_names-methods">taxa_names</a></span>(microbe)<span class="op">==</span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/taxa_names-methods">taxa_names</a></span>(microbe)[<span class="op">!</span>keep_ix][<span class="dv">1</span>])
nms[rnm] &lt;-<span class="st"> "other"</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/taxa_names-methods">taxa_names</a></span>(microbe) &lt;-<span class="st"> </span>nms
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rm">rm</a></span>(nms, rnm)

<span class="co"># bit of preprocessing </span>
metab &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(<span class="dv">1</span> <span class="op">+</span><span class="st"> </span>metab)</code></pre></div>
<p>Now I am going to do just a bit of processing to get data into a format for orthus. Note I have no extra metadata so we are just going to use an intercept in our model at this time.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Y &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/otu_table-methods">otu_table</a></span>(microbe, <span class="dt">taxa_are_rows=</span><span class="ot">TRUE</span>)
Z &lt;-<span class="st"> </span>metab <span class="co">#(metabolites are rows)</span>
X &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="kw"><a href="../reference/access_dims.html">nsamples</a></span>(microbe))

<span class="co"># save dims for easy reference</span>
N &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(Y)
P &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(Z)
Q &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(X)
D &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(Y)</code></pre></div>
<p>Now I am going to set up the priors. My priors are going to be similar to that of <em>pibble</em> but now we need to think about a prior for the covariance among the metabolites and between the metabolites and the log-ratios of the taxa. Remember, that priors must be defined in the <span class="math inline">\(ALR_D\)</span> (e.g., ALR with the reference being the D-th taxa; this may be changed in the future to make specifying priors more user friendly).</p>
<p>I am going to form our prior for <span class="math inline">\(\Sigma\)</span> by specifying <span class="math inline">\(\upsilon\)</span> and <span class="math inline">\(\Xi\)</span>. I will specify that I have weak prior beleif that the taxa are independent in terms of their log absolute abundance. We can translate this statement about covariance of log absolute abundance into a statement about log-ratio covariance by pre- and post-multiplying by the <span class="math inline">\(ALR_D\)</span> contrast matrix (which I refer to as <span class="math inline">\(GG\)</span> below). Additionally, I believe that there is likely no substantial covariance between the taxa and the metabolites and I assume the metabolites are likely indpednent.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">upsilon &lt;-<span class="st"> </span>(D<span class="op">-</span><span class="dv">1</span><span class="op">+</span>P)<span class="op">+</span><span class="dv">10</span> <span class="co"># weak-ish prior on covariance over joint taxa and metabolites</span>
Xi &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(D<span class="op">-</span><span class="dv">1</span><span class="op">+</span>P)
GG &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(D<span class="op">-</span><span class="dv">1</span>), <span class="op">-</span><span class="dv">1</span>)
Xi[<span class="dv">1</span><span class="op">:</span>(D<span class="op">-</span><span class="dv">1</span>), <span class="dv">1</span><span class="op">:</span>(D<span class="op">-</span><span class="dv">1</span>)] &lt;-<span class="st"> </span>GG<span class="op">%*%</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(D) <span class="op">%*%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/transpose-methods">t</a></span>(GG)
Xi &lt;-<span class="st"> </span>Xi <span class="op">*</span><span class="st"> </span>(upsilon<span class="op">-</span>D<span class="op">-</span>P) <span class="co"># this scales Xi to have the proper mean we wanted</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/image">image</a></span>(Xi)</code></pre></div>
<p><img src="orthus_files/figure-html/unnamed-chunk-5-1.png" width="480"></p>
<p>Note the stucture of this prior, everything is independent but there is a moderate positive covariance between the log-ratios based on their shared definition in terms of the <span class="math inline">\(D\)</span>-th taxa.</p>
<p>The other parts of the prior are less interesting. We are going to state that our mean for <span class="math inline">\(\Lambda\)</span> is centered about <span class="math inline">\(\mathbf{0}\)</span> and that the signal-to-noise ratio in the data is approximately 1 (this later part is specified by <span class="math inline">\(\Gamma=I\)</span>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Gamma &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(Q)
Theta &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dv">0</span>, D<span class="op">-</span><span class="dv">1</span><span class="op">+</span>P, <span class="dv">1</span>)</code></pre></div>
<p>Finally I fit the model.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/orthus_fit.html">orthus</a></span>(Y, Z, X, <span class="dt">Theta=</span>Theta, <span class="dt">Gamma=</span>Gamma, <span class="dt">Xi=</span>Xi, <span class="dt">upsilon=</span>upsilon, <span class="dt">n_samples=</span><span class="dv">1000</span>)</code></pre></div>
<p>Next we are going to transform the log-ratios from <span class="math inline">\(ALR_D\)</span> to the <span class="math inline">\(CLR\)</span>. I have written all the transformation functions, <em>e.g.</em>, <code>to_clr</code> etc… to work on <code>orthusfit</code> objects in a similar manner to how they work on <code>pibblefit</code> objects. For <code>orthusfit</code> objects they only transform the log-ratio components of paramters leaving the other parts of inferred model parameters (<em>i.e.</em>, the parts associated with the metabolites) untouched.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/stray_transforms.html">to_clr</a></span>(fit)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(fit)
<span class="co">#&gt; orthusfit Object: </span>
<span class="co">#&gt;   Number of Samples:      12 </span>
<span class="co">#&gt;   Number of Categories:       114 </span>
<span class="co">#&gt;   Number of Zdimensions:  405 </span>
<span class="co">#&gt;   Number of Covariates:       1 </span>
<span class="co">#&gt;   Number of Posterior Samples:    1000 </span>
<span class="co">#&gt;   Contains Samples of Parameters:Eta  Lambda  Sigma</span>
<span class="co">#&gt;   Coordinate System:      clr</span></code></pre></div>
</div>
<div id="investigate-model-results" class="section level1">
<h1 class="hasAnchor">
<a href="#investigate-model-results" class="anchor"></a>Investigate Model Results</h1>
<p>There are a ton of ways to visualize the inferred model. I could make network diagrams relating taxa to taxa, taxa to metabolites and metabolites to metabolites. I could look at a low dimensional representation of joint covariance to create something very much akin to cannonical correlation analysis (CCA). I could look at how well the metabolites predict the taxa and vice-versa. But for the sake of simplicity I will do something much simpler. Here I am just going to find a list of taxa metabolite covariances that the model is very confident about.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># First just look ath the cross-covariances fit by the model</span>
<span class="co"># (covariance between taxa in CLR coordinates and metabolites)</span>
<span class="co"># This requires that we extract the corner of Sigma. </span>
xcor &lt;-<span class="st"> </span>fit<span class="op">$</span>Sigma[<span class="dv">1</span><span class="op">:</span>D, D<span class="op">:</span>(D<span class="op">-</span><span class="dv">1</span><span class="op">+</span>P),]

<span class="co"># Initial preprocessing to speed up computation of posterior intervals</span>
<span class="co"># As there are alot of cross-covariance terms we are going to first </span>
<span class="co"># weed down the list of things we have to look at by first pass </span>
<span class="co"># selecting only those taxa that have a large posterior mean for the covariance</span>
xcor.mean &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(xcor, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>), mean)
to.analyze &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/driver/topics/gather_array">gather_array</a></span>(xcor.mean, cov, taxa, metabolite) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(<span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(cov)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span>.[<span class="dv">1</span><span class="op">:</span><span class="dv">1000</span>,] <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">tm =</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(taxa, <span class="st">"_"</span>, metabolite))

<span class="co"># Subset Covariance to those we are interested in and calculate posterior </span>
<span class="co"># confidence intervals. </span>
xcor.summary &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/driver/topics/gather_array">gather_array</a></span>(xcor, cov, taxa, metabolite, iter) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">tm=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(taxa, <span class="st">"_"</span>, metabolite)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(tm <span class="op">%in%</span><span class="st"> </span>to.analyze<span class="op">$</span>tm) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">taxa =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(Y)[taxa], <span class="dt">metabolite =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(Z)[metabolite]) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(taxa, metabolite) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/driver/topics/summarise_posterior">summarise_posterior</a></span>(cov) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(mean) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(taxa <span class="op">!=</span><span class="st"> 'other'</span>) <span class="co"># we don't care about these</span>

<span class="co"># Select those covariances where the model has high certainty (95%) that</span>
<span class="co"># the true covariance is not zero. </span>
xcor.summary <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sign">sign</a></span>(p2.<span class="dv">5</span>)<span class="op">==</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sign">sign</a></span>(p97.<span class="dv">5</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(mean) <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>)
<span class="co">#&gt; # A tibble: 203 x 8</span>
<span class="co">#&gt; # Groups:   taxa [15]</span>
<span class="co">#&gt;    taxa  metabolite   p2.5   p25   p50  mean   p75  p97.5</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;       &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt;  1 722   206.0445922 -6.27 -3.92 -3.09 -3.27 -2.38 -1.52 </span>
<span class="co">#&gt;  2 7816  206.0445922 -5.48 -3.16 -2.44 -2.63 -1.83 -1.06 </span>
<span class="co">#&gt;  3 722   290.9298419 -4.85 -3.09 -2.43 -2.58 -1.93 -1.18 </span>
<span class="co">#&gt;  4 18182 380.1846197 -5.06 -3.02 -2.34 -2.52 -1.80 -1.02 </span>
<span class="co">#&gt;  5 722   181.4504354 -4.57 -3.01 -2.30 -2.49 -1.83 -1.07 </span>
<span class="co">#&gt;  6 722   177.0565368 -4.60 -2.97 -2.32 -2.47 -1.85 -1.08 </span>
<span class="co">#&gt;  7 19517 380.1846197 -5.16 -3.01 -2.30 -2.46 -1.73 -0.927</span>
<span class="co">#&gt;  8 722   176.0343919 -4.60 -2.93 -2.30 -2.45 -1.83 -1.09 </span>
<span class="co">#&gt;  9 722   180.072273  -4.69 -2.94 -2.26 -2.42 -1.74 -1.01 </span>
<span class="co">#&gt; 10 2943  380.1846197 -5.10 -2.88 -2.25 -2.41 -1.73 -0.981</span>
<span class="co">#&gt; # … with 193 more rows</span></code></pre></div>
<p>So it looks there there are a few hundred covariances that we can be fairly confident about.</p>
</div>
<div id="qualifications-and-caution" class="section level1">
<h1 class="hasAnchor">
<a href="#qualifications-and-caution" class="anchor"></a>Qualifications and Caution</h1>
<p>Please note, I performed this analysis to demonstrate the use of <em>orthus</em> which is a model that I have been repeatedly asked for. I think its a cool model and could be quite useful in the right circumstances. But I would like to point out a few philosophical points about the analysis I performed above.</p>
<p>First, I performed this analysis just to demonstrate <em>orthus</em>. I really don’t know the data showcased here. What is metabolite <code>206.0445922</code>? I have no idea. For some reason this is how the metabolites in that dataset were named. For the same reason I have left the taxa indexed by sequence variant number.</p>
<p>Second (and more important), identifying relationships between taxa and metabolites (or between any two high-dimensinal multivariate datasets) is really difficult! Here we are looking at just 114 taxa and 405 but this leads to 46170 possible covariances and here we only have 12 samples! Yes <em>orthus</em> is a Bayesian model, and Yes, Bayesian models can be quite useful when there are more parameters than samples, but there is a limit of reasonability. Really, Bayesian models are great when you can perfectly capture your prior beleifs with your prior. But how often can that really be done perfectly? As such I would caution users, don’t run out and try applying <em>orthus</em> on every possible dataset with massive dimensions. Think carefully about what taxa or metabolites (or whatever else you are looking at) you really care about and analyze those. Use a bit of expert knowledge to guide your science.</p>
<p>That said, please do not take the above comment as a particular limitation of <em>orthus</em>. It’s not that some other model will overcome this issue. It’s a fundamental issue with answering a very very difficult question (e.g., the covariance between high-dimensional datasets) using limited data.</p>
<p>Alright, thats probably enough philosophising for an R package Vignette. I hope you enjoy <em>orthus</em>.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-callahan2016">
<p>Callahan, Ben J, Kris Sankaran, Julia A Fukuyama, Paul J McMurdie, and Susan P Holmes. 2016. “Bioconductor Workflow for Microbiome Data Analysis: From Raw Reads to Community Analyses.” <em>F1000Research</em> 5. Faculty of 1000 Ltd.</p>
</div>
<div id="ref-kashyap2013">
<p>Kashyap, Purna C, Angela Marcobal, Luke K Ursell, Samuel A Smits, Erica D Sonnenburg, Elizabeth K Costello, Steven K Higginbottom, et al. 2013. “Genetically Dictated Change in Host Mucus Carbohydrate Landscape Exerts a Diet-Dependent Effect on the Gut Microbiota.” <em>Proceedings of the National Academy of Sciences</em> 110 (42). National Acad Sciences: 17059–64.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#strayorthus-the-model">stray::orthus, the model</a></li>
      <li><a href="#joint-modeling-of-microbial-16s-data-and-metabolomics">Joint modeling of Microbial 16S data and Metabolomics</a></li>
      <li><a href="#investigate-model-results">Investigate Model Results</a></li>
      <li><a href="#qualifications-and-caution">Qualifications and Caution</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Justin Silverman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
